packages/ra-core/src/form/FormDataConsumer.js

```diff
@@ -42,8 +43,58 @@ const REDUX_FORM_NAME = 'record-form';
  *     </Edit>
  * );
  */
-const FormDataConsumer = ({ children, formData, ...rest }) => {
-    const ret = children({ formData, ...rest });
+const FormDataConsumer = ({ children, formData, source, ...rest }) => {
+    let scopedFormData = formData;
+    let getSource;
+    let getSourceHasBeenCalled = false;
+
+    // If we have a source, we are in an ArrayInput component
+    if (source) {
+        const matches = ARRAY_SOURCE_REGEX.exec(source);
+
+        if (matches.length > 2) {
+            const [, arraySource, index] = matches;
+            scopedFormData = formData[arraySource][parseInt(index)];
+            getSource = scopedSource => {
+                getSourceHasBeenCalled = true;
+                return `${arraySource}[${index}].${scopedSource}`;
+            };
+        }
+    }
+    const ret = children({ formData, scopedFormData, getSource, ...rest });
+
+    if (
+        ret &&
+        !getSourceHasBeenCalled &&
+        process.env.NODE_ENV !== 'production'
+    ) {
+        console.warn( // eslint-disable-line
+            `You have an input inside a FormDataConsumer inside an ArrayInput and you did not called the getSource function supplied by the FormDataConsumer component. This is required for your input to get the proper source and name.
+
+            <ArrayInput source="users">
```

Comment:
the auto formatting added *a lot* of leading spaces on each line, which I don't think will look good on the console.

Reply:
It was fine but I extracted it anyway to be sure :)

![screen](http://pix.toile-libre.org/upload/original/1533484303.png)

I indented the code on purpose but it may be not so nice to copy/paste it with this indentation. What do you think?